newspaper/mthreading.py

```diff
@@ -81,28 +80,29 @@ def __init__(self, config=None):
         >>> cnn_paper.articles[50].html
         u'<html>blahblah ... '
         """
-        self.papers = []
+        self.papers = {}
         self.pool = None
         self.config = config or Configuration()
 
     def join(self):
         """
-        Runs the mtheading and returns when all threads have joined
-        resets the task.
+        Waits for all submitted papers to complete downloading their articles.
         """
         if self.pool is None:
             print('Call set(..) with a list of source '
                   'objects before .join(..)')
             raise
-        self.pool.wait_completion()
-        self.papers = []
-        self.pool = None
 
-    def set(self, paper_list, threads_per_source=1):
-        self.papers = paper_list
-        num_threads = threads_per_source * len(self.papers)
-        timeout = self.config.thread_timeout_seconds
-        self.pool = ThreadPool(num_threads, timeout)
+        # Wait for completion of all futures.
+        self.pool.wait_completion()
 
-        for paper in self.papers:
-            self.pool.add_task(paper.download_articles)
+    def set(self, paper_list, num_threads=8, fun=perform_download_articles):
```

Comment:
This changes control behavior, it used to be 1 thread per source but with this change it will be 8 threads no matter what. I'm scared existing callers using this API will break after this change:

Can we revert this bit to go back to `threads_per_source`, later I can add another API to allow num_threads to be configured because we are moving to a world where `set` can be called on Source objects and also Article objects

Reply:
I understand the concert, but I think that the number of threads shouldn't effect callers that used to use the old version. The behavior of the old pool only let the caller wait until all the sources were done, so how they get to that point shouldn't matter. I'm still going to make the change to make the API the same as it used to be though.

Reply:
👌